src/core/Akka/Actor/ActorProducerPipeline.cs

```diff
@@ -0,0 +1,251 @@
+    {
+        /// <summary>
+        /// Determines if current plugin can be applied to provided <paramref name="actor"/> instance.
+        /// </summary>
+        bool CanBeAppliedTo(ActorBase actor);
+
+        /// <summary>
+        /// Plugin behavior applied to <paramref name="actor"/> instance when the new one is being created.
+        /// </summary>
+        void AfterActorCreated(ActorBase actor, IActorContext context);
+
+        /// <summary>
+        /// Plugin behavior applied to <paramref name="actor"/> instance before the actor is being recycled.
+        /// </summary>
+        void BeforeActorTerminated(ActorBase actor, IActorContext context);
+    }
+
+    /// <summary>
+    /// Base actor producer pipeline plugin class.
+    /// </summary>
+    public abstract class ActorProducerPluginBase : IActorProducerPlugin
+    {
+        /// <summary>
+        /// By default derivatives of this plugin will be applied to all actors.
+        /// </summary>
+        public virtual bool CanBeAppliedTo(ActorBase actor)
+        {
+            return true;
+        }
+
+        /// <summary>
+        /// Plugin behavior applied to <paramref name="actor"/> instance when the new one is being created.
+        /// </summary>
+        public virtual void AfterActorCreated(ActorBase actor, IActorContext context) { }
+
+        /// <summary>
+        /// Plugin behavior applied to <paramref name="actor"/> instance before the actor is being recycled.
+        /// </summary>
+        public virtual void BeforeActorTerminated(ActorBase actor, IActorContext context) { }
+    }
+
+    /// <summary>
+    /// Base generic actor producer pipeline plugin class.
+    /// </summary>
+    public abstract class ActorProducerPluginBase<TActor> : IActorProducerPlugin where TActor : ActorBase
+    {
+        /// <summary>
+        /// By default derivatives of this plugin will be applied to all actors inheriting from <typeparam name="TActor">actor generic type</typeparam>.
+        /// </summary>
+        public virtual bool CanBeAppliedTo(ActorBase actor)
+        {
+            return actor is TActor;
+        }
+
+        void IActorProducerPlugin.AfterActorCreated(ActorBase actor, IActorContext context)
+        {
+            AfterActorCreated(actor as TActor, context);
+        }
+
+        void IActorProducerPlugin.BeforeActorTerminated(ActorBase actor, IActorContext context)
+        {
+            BeforeActorDestroyed(actor as TActor, context);
+        }
+
+        /// <summary>
+        /// Plugin behavior applied to <paramref name="actor"/> instance when the new one is being created.
+        /// </summary>
+        public virtual void AfterActorCreated(TActor actor, IActorContext context) { }
+
+        /// <summary>
+        /// Plugin behavior applied to <paramref name="actor"/> instance before the actor is being recycled.
+        /// </summary>
+        public virtual void BeforeActorDestroyed(TActor actor, IActorContext context) { }
```

Comment:
I don't think "Destroy" is used anywhere else in Akka.NET afaik, is this name clear enough for users to understand that it applies to when the instance is about to be killed?
Not saying that it's a bad name, just so we are consistent and clear with naming.